### PR TITLE
T5871: ipsec remote access VPN: specify "cacerts" for client auth

### PR DESCRIPTION
**Pull request by lucasec**
## Change Summary

For authentication methods that depend on validating a client certificate against a CA (e.g. EAP-TLS), we currently do not explicitly tell strongswan which CA to use. This PR specifies the `cacerts` option explicitly for every connection to ensure the connection only accepts certificates signed by its specific CA.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5871

## Related PR(s)

N/A

## Component(s) name

ipsec remote-access

## Proposed changes

The configured CA certificate fro the connection is added to the swanctl.conf connection definition using the `cacerts` option.

## How to test

Example configuration:
```
[edit vpn ipsec remote-access connection ClientVPN]
lucas@lcn-router# show
 authentication {
     client-mode eap-tls
     local-id <local id>
     server-mode x509
     x509 {
         ca-certificate <ca certificate name>
         certificate <server certificate name>
     }
 }
 local-address <router IP address>
 ...
```

Validate that `/etc/swanctl/swanctl.conf` specifies `cacerts = <ca certificate name>_1.pem` under `remote`.

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: data/templates/ipsec/swanctl/remote_access.j2 [hunk 1]
```diff
@@ -32,6 +32,9 @@
             auth = {{ rw_conf.authentication.client_mode }}
 {% if rw_conf.authentication.client_mode.startswith("eap") %}
             eap_id = %any
+{% endif %}
+{% if rw_conf.authentication.client_mode == 'eap_tls' or rw_conf.authentication.client_mode == 'x509' %}
```
Comment (c-po) [inline data/templates/ipsec/swanctl/remote_access.j2]: You should use `eap-tls` over `eap_tls`. Please rephrase it to:

`{% if rw_conf.authentication.client_mode is vyos_defined('eap-tls') or rw_conf.authentication.client_mode is vyos_defined('x509') %}`

Else the code path is not hit for eap-tls client auth.

**Review by c-po — CHANGES_REQUESTED** [review]

File: data/templates/ipsec/swanctl/remote_access.j2 (hunk 1 above)
Comment (c-po) [reply data/templates/ipsec/swanctl/remote_access.j2]: Please also update the IPSec remote-access smoketest (test_07) in https://github.com/vyos/vyos-1x/blob/current/smoketest/scripts/cli/test_vpn_ipsec.py#L551 to validate this option.

**Comment by lucasec** [comment]
@c-po Updated with smoketests and rebase onto latest.

@sarthurdev I did a little more investigation into the CA chain behavior you introduced in https://github.com/vyos/vyos-1x/commit/1ac230548c86d3308ff5b479b79b0e64b75a0e8a. From my understanding, the user would specify the intermediate CA name under `authentication x509 ca-certificate`, and then the config logic would render the intermediate as `CA_name_1.pem`, next parent as `CA_name_2.pem`, up to the root being `CA_name_N.pem`.

I would think in most cases the user would want strongswan to be "strict", as in only accept client certs signed by the intermediate. In that case, the current behavior in this PR, which renders `cacerts = CA_name_1.pem`, would match the user's expectation. But I could see valid cases where the user would want more broad acceptance, e.g. accept client certs signed by other intermediates under the same root (closer to the current behavior without this PR).

Definitely open to more discussion on this, e.g. are we overloading too much behavior on the `x509 ca-certificate` option.

**Comment by sarthurdev** [comment]
> @sarthurdev I did a little more investigation into the CA chain behavior you introduced in [1ac2305](https://github.com/vyos/vyos-1x/commit/1ac230548c86d3308ff5b479b79b0e64b75a0e8a). From my understanding, the user would specify the intermediate CA name under `authentication x509 ca-certificate`, and then the config logic would render the intermediate as `CA_name_1.pem`, next parent as `CA_name_2.pem`, up to the root being `CA_name_N.pem`.
> 
> I would think in most cases the user would want strongswan to be "strict", as in only accept client certs signed by the intermediate. In that case, the current behavior in this PR, which renders `cacerts = CA_name_1.pem`, would match the user's expectation. But I could see valid cases where the user would want more broad acceptance, e.g. accept client certs signed by other intermediates under the same root (closer to the current behavior without this PR).

Thanks for raising this, it's a very good point. Thinking my commit should be reverted and a multi value `ca-certificate` node is made available instead.

**Comment by lucasec** [comment]
If I understand correctly, there are sort of two behaviors an admin might want to accomplish by supplying an intermediate to strongswan:

1. Allow strongswan to "complete the chain" if the client supplies a cert signed by the intermediate but does not supply the intermediates along with it (I am familiar with transmitting multiple certs over the wire in the context of HTTP, but not as much with if/how that is implemented in IKE).
2. The admin might want to only have certs issued by a specific intermediate trusted.

In case 1, if there were other intermediates below the root and the admin also wanted those trusted, they would need to be able to load all of the different intermediates at the same level into strongswan. If there are multiple levels of intermediates it seems like the number of certs that need to be loaded could explode out quite a bit.

And in case 2, if you only want the intermediate trusted, would strongswan need to have access to the root at all?

File: data/templates/ipsec/swanctl/remote_access.j2 [hunk 2]
```diff
@@ -34,6 +34,11 @@
 {% elif rw_conf.authentication.client_mode.startswith("eap") %}
             auth = {{ rw_conf.authentication.client_mode }}
             eap_id = %any
+{% endif %}
+{% if rw_conf.authentication.client_mode is vyos_defined('eap-tls') or rw_conf.authentication.client_mode is vyos_defined('x509') %}
+{#          in the case of an intermediate/root chain, assume user wants us to valiadate against #}
+{#          the lowest specified cert in the chain (e.g. the intermediate) #}
+            cacerts = {{ rw_conf.authentication.x509.ca_certificate }}_1.pem
```
Comment (sarthurdev) [inline data/templates/ipsec/swanctl/remote_access.j2]: Need to change to:
`cacerts = {{ '.pem,'.join(rw_conf.authentication.x509.ca_certificate) ~ '.pem' }}`

(Should be correct, though will need testing)

File: smoketest/scripts/cli/test_vpn_ipsec.py [hunk 3]
```diff
@@ -754,6 +754,7 @@ def test_08_ikev2_road_warrior_client_auth_eap_tls(self):
             f'id = "{local_id}"',
             f'auth = pubkey',
             f'certs = peer1.pem',
+            f'cacerts = MyVyOS-CA_1.pem',
```
Comment (sarthurdev) [inline smoketest/scripts/cli/test_vpn_ipsec.py]: `_1` needs to be dropped

File: smoketest/scripts/cli/test_vpn_ipsec.py [hunk 4]
```diff
@@ -864,6 +865,7 @@ def test_09_ikev2_road_warrior_client_auth_x509(self):
...
```
Comment (sarthurdev) [inline smoketest/scripts/cli/test_vpn_ipsec.py]: `_1` needs to be dropped

**Review by sarthurdev — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
@lucasec can you rebase your work to the latest `current` now that #3205 is merged?

**Comment by lucasec** [comment]
Rebased accordingly. Verified the updated smoketests pass but I am holding testing this on my live system until I finish up some other changes tomorrow (feel free to merge or hold at your discretion).

File: data/templates/ipsec/swanctl/remote_access.j2 (hunk 2 above)
Comment (lucasec) [reply data/templates/ipsec/swanctl/remote_access.j2]: That syntax should work but I ended up taking what I thought was a "cleaner" approach of adding a filter. You are welcome to veto that though.

**Comment by lucasec** [comment]
Installed on my live system and looks fine.

I am still trying to chase some other issue causing system crashes seemingly related to ipsec, but those predate this PR (I am now investigating whether they are possibly related to https://github.com/vyos/vyos-1x/pull/3157).

**Review by sarthurdev — APPROVED** [review]

**Comment by sever-sever** [comment]
@lucasec is it ready to merge?

**Comment by lucasec** [comment]
Yes—this one is clear to merge.

**Review by sever-sever — APPROVED** [review]

File: python/vyos/template.py [hunk 5]
```diff
@@ -172,6 +172,14 @@ def force_to_list(value):
     else:
         return [value]
 
+@register_filter('prepend_str')
+def prepend_str(value, prefix):
```
Comment (c-po) [inline python/vyos/template.py]: Why not use the Jinja2 ~ operator to concat strings?

**Review by c-po — CHANGES_REQUESTED** [review]

File: python/vyos/template.py (hunk 5 above)
Comment (lucasec) [reply python/vyos/template.py]: We could.

I thought this was cleaner and would be easier to maintain than @sarthurdev’s suggestion here:
> Need to change to: `cacerts = {{ '.pem,'.join(rw_conf.authentication.x509.ca_certificate) ~ '.pem' }}`

(in particular, the use of `join` with `.pem` as the separator is clever, but IMHO not immediately obvious what it’s doing)

Comment (c-po) [reply python/vyos/template.py]: Please proceed with join and pem separator and add a proper comment. Then I'm happy to merge this one!

Comment (lucasec) [reply python/vyos/template.py]: Updated and confirmed smoke tests still pass.

**Review by c-po — APPROVED** [review]

**Comment by c-po** [comment]
@MegifyIo backport sagitta

**Comment by GurliGebis** [comment]
@c-po it looks like it didn't create the backport pr

**Comment by sarthurdev** [comment]
@mergifyio backport sagitta
